gammapy/scripts/jupyter.py

```diff
@@ -0,0 +1,149 @@
+                        semicolon = 1
+                    fmt = format_str(src_contents=fmt,
+                                     line_length=79).rstrip()
+                    if semicolon:
+                        fmt += ';'
+                except Exception as ex:
+                    logging.info(ex)
+                fmt = fmt.replace("###-MAGIC COMMAND-", "")
+            nb.cells[cellnumber]['source'] = fmt
+
+        # write formatted notebook
+        nbformat.write(nb, str(notebook))
+
+        # inform
+        print('Jupyter notebook {} painted in black.'.format(str(notebook)))
+
+
+@click.command(name='stripout')
+@click.pass_context
+def cli_jupyter_stripout(ctx):
+    """Strip output cells."""
+
+    nblist = build_nblist(ctx)
+
+    for notebook in nblist:
+
+        try:
+            subprocess.call("nbstripout '{}'".format(notebook), shell=True)
+            print('Jupyter notebook {} stripped out.'.format(str(notebook)))
+        except Exception as ex:
+            log.error('Error stripping file {}'.format(str(notebook)))
+            log.error(ex)
+
+
+@click.command(name='execute')
+@click.pass_context
+def cli_jupyter_execute(ctx):
+    """Execute jupyter notebook."""
+
+    nblist = build_nblist(ctx)
+
+    for notebook in nblist:
+
+        try:
+            t = time.time()
+            subprocess.call(
+                "jupyter nbconvert --allow-errors --ExecutePreprocessor.timeout=None --ExecutePreprocessor.kernel_name=python3 --to notebook --execute '{}'.format(notebook) --inplace",
+                shell=True)
+            t = (time.time() - t) / 60
+            print('Executing duration: {:.2f} mn'.format(t))
+        except Exception as ex:
+            log.error('Error executing file {}'.format(str(notebook)))
+            log.error(ex)
+
+
+@click.command(name='test')
+@click.pass_context
+def cli_jupyter_test(ctx):
+    """Check if jupyter notebook is broken."""
+
+    notebook = Path(ctx.obj['file'])
+    folder = Path(ctx.obj['fold'])
+    ignorelist = []
+
+    if ctx.obj['file']:
+        # ignore all files except jupyterfile
+        for f in notebook.parent.iterdir():
+            if notebook.name != f.name and f.name.endswith('.ipynb'):
+                nbname = f.name.replace('.ipynb', '')
+                ignorelist.append(nbname)
+        folder = notebook.parent
+
+        print('oups')
+
+    testnb = testipynb.TestNotebooks(
+        directory=str(folder), ignore=ignorelist)
+    TestCase.assertTrue(testnb, testnb.run_tests())
+
+
+def comment_magics(input):
+    """Comment magic commands when formatting cells."""
+
+    lines = input.splitlines(True)
+    output = ""
+    for line in lines:
+        new_line = ""
+        if line.startswith("%") or line.startswith("!"):
+            new_line = new_line + "###-MAGIC COMMAND-" + line
+        if new_line:
+            output = output + new_line
+        else:
+            output = output + line
+    return output
+
+
+def build_nblist(ctx):
```

Comment:
Suggest to make this a generator function that `yield`s the paths. That means you can remove the lines that make the empty list and return it at the end, and the `nblist.append` calls become `yield` statements.
You can then use it as an iterable in the for loop directly, or call `list()` on the generator if you need a list, but i nthis case you don't need a list.

Reply:
Done.